### ISSUE TITLE
Can the generated swagger definition support adding header for Basic Auth?

### ISSUE THREAD

**Issue by bbarman4u** [stoicflame/enunciate#468]
Currently many of my rest services are protected and can only be accessed for try out only if a basic auth header with the necessary user name and password is passed. Now many tools like Postman or Soap UI also provide the capability to inject the basic Auth. Also Swagger seems to provide it in their specification. Is there a configuration in the enunciate.xml that I could use to specify this ?
Link for reference -

https://github.com/swagger-api/swagger-ui#custom-header-parameters---for-basic-auth-etc

**Comment by stoicflame** [comment]
We could add an annotation, but I'm not sure what it should _do_. Are you looking for something like what was being asked at #368?

**Comment by bbarman4u** [comment]
Hi Stoicflame,

No this is not the issue that I was referencing. By default the generated swagger docs provides the option to pass the api_key at the top of the page which is populated in the query parameter. While this may not suit systems like mine where we have to pass headers with http basic auth to authenticate the calling rest service. While I am aware that I can hand edit the generated swagger files to add the capability in there, but it would have to be done every time there is a build which would make this exercise pointless.
Hence I wanted to know if enunciate has the capability may be in future version to provide options to pass headers with basic auth or may be headers for oauth, etc as part of the build by configurations with enunciate.xml

**Comment by stoicflame** [comment]
Okay, I can look into support for that. Will you give me specifics on how you hand-edit the swagger files to add the capability?

**Comment by bbarman4u** [comment]
Hi,
Let me start by saying that I am not an expert on swagger nor HTML, hence my solution was not all that elegant but rather a working solution only. So this is what I did -
1. Edited the index.html file under ui folder to add another input element for the username and password to be entered in the format "username:password" that I would then read using Jqeuery, convert it into base 64 and set a header with it. Ideally may be adding another drop down select box with different authentication options and based on the selection providing an input text area would have been better, but I was in a rush. Anyway details below -

- Edited the ui/index.html and added a new function "addBasicAuthInHeader()" that I added after the "addApiKeyAuthorization()".
- Added a new input text box as below -
`<div class='input'><input placeholder="username:password" id="input_basicAuthKey" name="authKey" type="text"/></div>`
- Here is my function for adding the basic auth in header -
```
	   function addBasicAuthInHeader(){
        var key = $('#input_basicAuthKey')[0].value;
        if(key && key.trim() != "") {
        	//Encode the string into base 64 format
			var base64EncodedKey = btoa(key);
			var basicAuthKey = "Basic " + base64EncodedKey;
			log("basicAuthKey ===" + basicAuthKey);
			//Add a basic Auth header, note the key needs to be unique per header
			window.swaggerUi.api.clientAuthorizations.add("basicAuthKey", new SwaggerClient.ApiKeyAuthorization("Authorization", basicAuthKey, "header"));
            log("Added Header " + basicAuthKey);
        }
      }
```
- Triggered the basic auth function anytime user entered something in the text box.
`$('#input_basicAuthKey').change(addBasicAuthInHeader);   `  

- Then patched the lib/enunciate-swagger-2.5.0.jar in my work space with this updated index.html to persist the change across my different projects any time I would regenerate the swagger docs.

- Attached is the index.html I edited in case my comments don't make it due to the code.
[index.html.txt](https://github.com/stoicflame/enunciate/files/393697/index.html.txt)

**Comment by stoicflame** [comment]
Thanks for sharing.

After taking a look at the solution, I'm going to pass on adding it to Enunciate. Users can configure Enunciate with a custom swagger "base" that Enunciate will use instead of the default one that contains the index.html. The custom base can contain the functionality as described above.

You can provide a custom base as described in the [Swagger Module Documentation](https://github.com/stoicflame/enunciate/wiki/Module-Swagger#updating-or-configuring-swagger-ui) and #304 and #466.
